### ISSUE TITLE
[Request] slow to refresh widgets on kodi start (delay refresh on startup)

### ISSUE THREAD

**Issue by djay** [drinfernoo/plugin.program.autowidget#53]
# Problem
If you have more than a few plugin based widgets refreshing can take a long time, esp on lower end hardware. Often the lists don't change that much so an outdated list might be fine temporarily. Some low end systems like android kill kodi in the background make restarts frequent and waiting for widgets to reload is annoying

# Solutions
- retain the cached widget data instead of cleaning them automatically on startup. 
- Refresh the widget display only after each widget data has been refreshed.
- Potentially delay the refresh a few minutes until after the kodi has fully started so refreshing doesn't slow down using kodi.
- potentially refreshing different widgets at different rates based on history of changes in the data.

**Comment by drinfernoo** [comment]
Unfortunately, there isn't much I can do here.

AutoWidget does no caching of its own, so any increases in caching performance would need to be handled in the "source" add-on.

Widgets already wait until new content has loaded before they change.

Delaying the refresh is definitely possible, and something I should probably be doing anyways.

Different widgets can be *set* to refresh at different rates, but since AutoWidget has no mechanism of knowing what items have changed, it would be very difficult to accomplish that.

**Comment by djay** [comment]
@drinfernoo I had a quick go at getting autowidget to cache the files in https://github.com/drinfernoo/plugin.program.autowidget/pull/55. Do you see any problems with this general approach?

**Comment by drinfernoo** [comment]
I think we can close this one now 👌

**Comment by djay** [comment]
@drinfernoo this isn't working to reduce startup times as much on my firestick as I'd hoped. Might need to reopen this.

I think whats happening is too many processes are being started at the same time and the new implementation with threads is not improving that. I suspect that the skinhelper is opening an autowidget process to update each widget and then each of those processes is creating a background thread which then kicks off another process for each plugin that needs updating. All at the same time as far too many plugins are starting their own background services.

My conclusions are
1. need to go back to having the autowidget background process doing the refreshes so its a single thread
2. need a better mechanism for the interprocess queue than the property I was using before. The property method was a bit dodgy as it had a race condition that could lose queued items. I'd either have to use the filesystem and create .queue files or try to implement some kind of lock property. Probably the former since order doesn't really matter.
3. Ideally it would be better if the autowidget service loaded last to reduce thrashing when the extra background refreshes are kicked off (ideas?). If not easy, then can simulate with a delay? Ideally though we want the widgets to background update as soon as possible because its frustrating for the user to been seeing yesterdays list when they know something new has been released.
4. Possibly sorting the paths in the queue before processing them might make things smoother.  It should load the same plugin in a row and prevent jumping between different plugins where multiple paths are from the same plugin.
4. This whole cache thing really needs to go into skinhelpers which would prevent any extra autowidget processes being created until later which would help even more. Unfortunately skinhelper code is terrible. I really tried to understand where this could be injected in that code but gave up.
5. I really need to work out how to test this code on my firestick. Python processes start way faster on notebook than on a lower end android so these issues don't become apparent in development. There are development releases based on whats merged to the develop branch that would make this easy @drinfernoo ?
6. Far too many addons have background services. I need to cut down on addons I don't use :)

**Comment by drinfernoo** [comment]
To be honest, this is essentially the biggest reason I had never gone after this before. In my opinion, it's best to have the *least* amount of background services as possible, but in most Kodi setups I've encountered, there are already a handful going *before any add-ons are even installed*. Kodi is unfortunately quite badly optimized for multiple threads, multiple processes, or really multiple of anything, at the same time 😅

3. I don't really know of a way to make it load last 🤦🏽‍♂️ The only thing that comes to mind is that it's *possible* that Kodi loads them in order by plugin ID, but that's a wild guess, and wouldn't really help us.

4. I don't know that it will make a difference. It would depend on whether or not the source addon is performing its own caching, whether it has `reuselanguageinvoker`, and if that tag is respected during JSON-RPC calls... The first two would be add-on dependent, and I don't know specifics on the latter.

5. Skin Helper is... dead, for all intents and purposes. As far as I know, it's not really being supported anymore into the future, and most of its functionalities have been superseded by other add-ons and modules.

6. `devrepo` releases are generated by a push to any branch (besides `master`, `repo`, `devrepo`, and `clean`). `repo` releases are generated by pushing a tag to `master`. This is handled by CircleCI, but it's been a long time since I set it up, and am not very familiar with the system. Likely the only way to have you be able to build to them, is for you be a contributor here, with push access.

7. 100% agreed. I only install add-ons I *use*, or that are *required* as dependencies. Many video (and other) add-ons run background services, and many of them even perform similar tasks, meaning that multiple add-ons running services can really pile on quickly.

In other news, the response from my testers so far has been positive, showing hefty speed increases in many cases. If you'd like to join us, come on in to https://discord.gg/upbQwMM, the waters great 😁

**Comment by djay** [comment]
@drinfernoo thats great news that its speeding up some systems (just not mine :( )
For testing what I'll do is resurrect my RP3. That will really test slow and is easy to to deploy dev code :)

if skin helper is dead then what are skins using to load manage? I realised development seemed dead which is another reason I didn't try and contribute to it, but didn't realise skins are switching to something else. It still seems widely used. I use aurora mod for example.

I will at least start with 1-2, which should have an impact. One thing I could do for 3 is that we could process the first queued item, get the skin to refresh it and wait until its then read again before going to the next item. That should mean things are settled down before everything but the first item is processed. Would just have to make sure the widget reload actually happens or have a timeout in case it doesn't. 

Tried discord but I think I need permission. userid is same as here.

**Comment by drinfernoo** [comment]
@djay For what it's worth, AuraMOD also hasn't been updated in quite some time.

A lot of skins still use Skin Helper (even ones that do still get updated), but that addon itself hasn't been updated significantly in almost a year (or more?), if I'm not mistaken.

An example of a skin that *doesn't* use Skin Helper is Arctic Horizon. It uses TMDbHelper for (I think) all of the features Skin Helper used to provide.

**Comment by djay** [comment]
@drinfernoo I've put disk queue idea into https://github.com/drinfernoo/plugin.program.autowidget/pull/60
